src/core/ddsc/src/dds_write.c

```diff
@@ -382,233 +410,188 @@ static bool fill_iox_chunk(dds_writer *wr, const void *sample, void *iox_chunk,
     // where we either have network readers (requiring serialization) or not.
 
     // do not serialize yet (may not need it if only using iceoryx or no readers)
-    d = ddsi_serdata_from_loaned_sample (ddsi_wr->type, writekey ? SDK_KEY : SDK_DATA, iox_chunk);
-    if(d == NULL) {
-      ret = DDS_RETCODE_BAD_PARAMETER;
-      goto release_chunk;
-    }   
-  } else {
+    d = (struct ddsi_serdata_iox *) ddsi_serdata_from_loaned_sample (ddsi_wr->type, writekey ? SDK_KEY : SDK_DATA, iox_chunk);
+  }
+  else
+  {
     // serialize for network since we will need to send via network anyway
     // we also need to serialize into an iceoryx chunk
- 
-    d = ddsi_serdata_from_sample (ddsi_wr->type, writekey ? SDK_KEY : SDK_DATA, data);
-    if(d == NULL) {
-      ret = DDS_RETCODE_BAD_PARAMETER;
-      goto release_chunk;
-    }
 
-    // Needed for the mixed case where serdata d was created for the network path
-    // but iceoryx can also be used.
-    // In this case d was created by ddsi_serdata_from_sample and we need
-    // to set the iceoryx chunk.
-    if(iceoryx_available) {
-      d->iox_chunk = iox_chunk;
-    } else {
-      d->iox_chunk = NULL;
-    }  
+    struct ddsi_serdata *dtmp = ddsi_serdata_from_sample (ddsi_wr->type, writekey ? SDK_KEY : SDK_DATA, data);
+    if (dtmp != NULL)
+    {
+      // Needed for the mixed case where serdata d was created for the network path
+      // but iceoryx can also be used.
+      // In this case d was created by ddsi_serdata_from_sample and we need
+      // to set the iceoryx chunk.
+      dtmp->iox_chunk = iox_chunk;
+      d = (struct ddsi_serdata_iox *) dtmp;
+    }
   }
+  if (d == NULL)
+  {
+    iox_pub_release_chunk (wr->m_iox_pub, iox_chunk);
+    return DDS_RETCODE_BAD_PARAMETER;
+  }
+  assert (d->x.iox_chunk != NULL);
 
   // refc(d) = 1 after successful construction
-
-  d->statusinfo = (((action & DDS_WR_DISPOSE_BIT) ? NN_STATUSINFO_DISPOSE : 0) |
-                  ((action & DDS_WR_UNREGISTER_BIT) ? NN_STATUSINFO_UNREGISTER : 0));
-  d->timestamp.v = tstamp;
+  set_statusinfo_timestamp ((struct ddsi_serdata_any *) d, tstamp, action);
 
   // 5. Deliver the data
-
   if(use_only_iceoryx) {
-    // deliver via iceoryx only
-    // TODO: can we avoid constructing d in this case?
-    if(deliver_data_via_iceoryx(wr, d)) {
-      ret = DDS_RETCODE_OK;
-    } else {
-      // Did not publish iox_chunk. We have to return the chunk (if any).      
-      iox_pub_release_chunk(wr->m_iox_pub, d->iox_chunk);
-      d->iox_chunk = NULL;
-      ret = DDS_RETCODE_ERROR;
+    // deliver via iceoryx and handle local non-iceoryx readers
+
+    // NB: there may be local non-iceoryx readers due to QoS like
+    // Keep Last not being supported with iceoryx, these are handled without
+    // iceoryx as otherwise they would get no data
+
+    if (deliver_locally_to_non_iox(ddsi_wr, d) == DDS_RETCODE_OK) {
```

Comment:
I think this is not the correct solution to the problem, because we first do our best to avoid touching the content if it came from a chunk (that is, Iceoryx owns the chunk), and then here we forward the pointer to subscriptions that Iceoryx is not aware of. I expect that in:
```
r1 = create reader via iceoryx (e.g., keep-last 1)
r2 = create reader not via iceoryx (e.g. keep-all)
w = create writer via iceoryx
x = borrow sample from w ; fill x ; write w, x
take r1
y = borrow sample from w ; fill y
take r2
```
r2 could see "y" where it should've seen "x". Perhaps my understanding of Iceoryx's buffer management is a little simplistic and it involves more work, but it seems like something of this sort could happen.

I propose changing the condition for `use_only_iceoryx` instead (actually also ensuring that it still means what the name suggests). That should be fairly easy: we just have to additionally require `wr->rdary.n_readers == 0`.

Reply:
Yes, I think you are right. It appears we need to restrict this further with this condition. I will change the latest commit.

Reply:
@eboasson Fixed, basically as you say. I was hoping to avoid the serialization of the mixed case we now use instead. In theory it would be possible but a headache.

Assuming we have all readers R = I + L with + disjoint union, I iceoryx readers, and L other local readers (basically those in the array). If I = IL + INL (iceoryx local and non-local). We could deliver the chunk as follows.

INL: chunk, via iceoryx, RHC will have a chunk
IL: chunk, via iceoryx, RHC will have a chunk
L: copy(!) chunk to RHC (instead of serialization and whatnot), RHC will have a copy

The IL case can be improved by an iceoryx change I will work on next:
move chunk directly to RHC (no copy, no iceoryx queuing etc.). But this is out of scope here.